urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/geom/twodim/SimplePolygon.java

```diff
@@ -0,0 +1,676 @@
+ * The interval representation is used only for internal computation. 
+ */
+public class SimplePolygon implements Polygon {
+	protected final Point[] points;
+	protected final CohoType type;
+	public CohoType type(){
+		return type;
+	}
+	//CONSIDER: Is it safe? It might be changed outside 
+	public SimplePolygon(Point[] points){
+		this(points,true);
+	}
+	public SimplePolygon(Collection<Point> points){
+		this(points.toArray(new Point[points.size()]));
+	}
+	public SimplePolygon(Collection<Point> points, boolean check){
+		this(points.toArray(new Point[points.size()]),check);
+	}
+	public SimplePolygon(SimplePolygon p){
+		this(p.points,false);
+	}
+	//TODO: not check?
+	public SimplePolygon(Iterator<Point> iter){
+		ArrayList<Point> p = new ArrayList<Point>();
+		type = p.get(0).type();
+		while(iter.hasNext()){
+			Point pt = iter.next();
+			if(type!=pt.type())
+				throw new GeomException("Type for polygon are not the same");
+			p.add(pt);
+		}
+		points = p.toArray(new Point[p.size()]);
+	}
+	//NOTICE: the points should be anti-clock-wise
+	//if check=false, we assign points directly, it's much faster and efficient.
+	//However, we should make sure 1)points is never changed outside 2)polygon is valid
+	protected SimplePolygon(Point[] points, boolean check){
+		if(!check){
+			this.points = points;//NOTE: points can't be changed outside. 
+			this.type = points[0].type();
+		}else{
+			//if(points==null || points.length==0){
+			if(points==null || points.length <1 ){
+				//throw new GeomException("null polygon");
+				this.points = new Point[0];
+				this.type = CohoDouble.type; //NOTE Let matlab handle it
+			}else{
+				//Use hashset rather than contain() of ArrayList to save time
+				HashSet<Point> set = new HashSet<Point>();
+				ArrayList<Point> list = new ArrayList<Point>();
+				this.type = points[0].type();
+				for(int i=0; i<points.length; i++){
+					if(set.add(points[i])){//copy and remove duplicated points
+						if(type!=points[i].type())
+							throw new GeomException("Type for polygon is not the same");
+						list.add(points[i]);
+					}else{
+						//System.out.println("repeated point "+points[i]+"found");
+					}
+				}
+				if(list.size()<3){
+					throw new GeomException("A polygon must have at least three vertices");
+				}
+				this.points = list.toArray(new Point[list.size()]);
+			}
+		}
+	}
+	
+	//NOTICE: the direction of segment must be anti-clock-wise.
+	public SimplePolygon(Segment[] segs){
+		Point curr;
+		try{
+			this.type = segs[0].type();
+			Hashtable<Point,Point> map  = new Hashtable<Point,Point>();
+			for(int i=0; i<segs.length; i++){
+				if(type!=segs[i].type())
+					throw new GeomException("type for polygon is not the same");
+				map.put(segs[i].p(0), segs[i].p(1));
+			}
+			points = new Point[segs.length];
+			curr = segs[0].p(0);
+			for(int i=0; i<segs.length;i++){
+				points[i] = curr;
+				curr = map.get(curr);
+			}
+		}catch(Exception e){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)